### PR TITLE
New tools documentation 

### PR DESCRIPTION
**Pull request by GFMoraga**
# Change Summary
Made a new `tools` directory in `docs/source/reference/tools` . This will have documentation on how to use the tools that are created. 

### PLEASE give feedback on .rst formatting and what to add/not add. 
- how do I do the index (tree?) attachment if I need to. 

This should check off a "to-do" in #75 

## New Files

- xtce_generator.rst
   - documentation on how to use the `xtce_generator_template.py` to create instrument specific scripts

### CODE REVIEW

File: imap_processing/codice/packet_definitions/P_COD_HI_INSTRUMENT_COUNTERS.xml [hunk 1]
```diff
@@ -3,42 +3,42 @@
 	<xtce:Header date="2023-09" version="1.0" author="IMAP SDC" />
```
Comment (maxinelasp) [inline imap_processing/codice/packet_definitions/P_COD_HI_INSTRUMENT_COUNTERS.xml]: These XML files should be in a different PR, they aren't related to the documentation.

File: docs/source/reference/tools/xtce_generator.rst [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+Prerequisites
+-------------
+
+Before you begin, ensure you have the following:
+
+- Python installed (version 3.9 or higher)
+- The required Python packages installed:
+    - pathlib
+    - tools.xtce_generation.telemetry_generator
```
Comment (greglucas) [inline docs/source/reference/tools/xtce_generator.rst]: pathlib is a part of the standard library, and I don't think you need to install telemetry_generator but rather you just use it right?

I'd suggest indicating how you do this with the poetry environments and linking to those documents as well. Something similar to this, feel free to reword as you want.

> Generating XTCEs is only done whenever packet definitions get updated, and thus it is not a part of the main processing package. To use it there are a few extra dependencies like `pandas` that you can install with `poetry install --with-extras xtce`.

File: docs/source/reference/tools/xtce_generator.rst [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+How to Use
+----------
+
+Define the instrument name in the `main()` function by setting the `instrument_name`
+variable to the name of your instrument.
+
+::
```
Comment (greglucas) [inline docs/source/reference/tools/xtce_generator.rst]:
```suggestion
variable to the name of your instrument::
```

I think you can eliminate this? Or alternatively you should add a `.. code::` or some qualifier, the double colons feel empty hanging out there otherwise. (Also to the ones below this)

File: docs/source/reference/tools/xtce_generator.rst [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+    module_path = f"{current_directory}/../../imap_processing"
+
+    # This is the path of the output directory
+    packet_definition_path = f"{module_path}/{instrument_name}/packet_definitions"
+
+    # This is the path to the excel file that contains the telemetry definitions
+    path_to_excel_file = f"{current_directory}/your_packet.xlsx"
+
+Define packet names and apIds. This is case sensitive. The packet names and apIds
```
Comment (greglucas) [inline docs/source/reference/tools/xtce_generator.rst]:
```suggestion
Define packet names and Application Process Identifiers (APIDs). This is case sensitive. The packet names and apIds
```

Maybe even make the APID a link? https://sanaregistry.org/r/space_packet_protocol_application_process_id/

Comment (greglucas) [inline docs/source/reference/tools/xtce_generator.rst]: What is case sensitive? The packet name?

Do the packet names need to be exactly what the tabs of the spreadsheet are? It would be helpful to clarify what this should be or if I am free to choose what I want.

File: docs/source/reference/tools/xtce_generator.rst [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+must match the names and apIds in the excel file. You can use as many packet names
+and apIds as you want. The apId should not be hexagon format. It should be an integer.
```
Comment (greglucas) [inline docs/source/reference/tools/xtce_generator.rst]:
```suggestion
and apIds as you want. The APID should be an integer (not hexadecimal).
```

File: docs/source/reference/tools/xtce_generator.rst [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        # ... (other packet definitions)
+    }
+
+
+Generating Telemetry XML Files
+-------------------------------
+
+This code block generates telemetry XML files for the packets defined in the
+`packets` dictionary. The telemetry XML files are saved in the directory specified
+by the `packet_definition_path` variable.
```
Comment (greglucas) [inline docs/source/reference/tools/xtce_generator.rst]:
```suggestion
Once you have your xtce processing file defined, you can run it with the following command:
    python xtce_generator_codice.py
```

I'd give the explicit command to use here after the file setup is complete. We don't care as much about what the code is doing (IMO), this is more of a user-guide for how to invoke the routines.

File: docs/source/reference/tools/xtce_generator.rst [hunk 7]
```diff
@@ -0,0 +1,81 @@
+.. _xtce_generator:
+
+Generating Telemetry XML with Python Script
+===========================================
+
+This document provides steps and information on how to use
+`xtce_generator_template.py` script as a base for users to generate
+telemetry XML files using the TelemetryGenerator class. The script is designed to
```
Comment (maxinelasp) [inline docs/source/reference/tools/xtce_generator.rst]:
```suggestion
telemetry XML files. The script is designed to
```

**Review by greglucas — COMMENTED** [review]
I'm not sure where we want to put this, but if we want to keep it in the API reference section.
https://github.com/IMAP-Science-Operations-Center/imap_processing/blob/b1a7b65be84de48c1a748155250ad2c4f3642910/docs/source/reference/index.rst?plain=1#L20
Adding another heading below instruments would be a good spot for it I think.

I am somewhat hesitant to put it in the API reference though because it isn't a part of the installed package. Maybe we want this to be up in the development section instead and then add it to the toc over there?

File: docs/source/reference/tools/xtce_generator.rst [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+An example of how to use the script is `xtce_generator_codice.py` which is also
+located in the `tools/xtce_generation` directory.
+
+Prerequisites
+-------------
+
+Before you begin, ensure you have the following:
+
+- Python installed (version 3.9 or higher)
+- The required Python packages installed:
```
Comment (maxinelasp) [inline docs/source/reference/tools/xtce_generator.rst]: Instead of saying "you need these installed" we should add pathlib to poetry and include instructions for installing that:
`poetry install --extras tools`

File: docs/source/reference/tools/xtce_generator.rst (hunk 5 above)
Comment (maxinelasp) [inline docs/source/reference/tools/xtce_generator.rst]:
```suggestion
Define packet names and apIds. This is case sensitive. You can use as many packet names
and apIds as you want. The apId should not be hexagon format. It should be an integer.
```

Remove the sentence about excel files, as other projects may have different ways of defining packets and it would be confusing

File: docs/source/reference/tools/xtce_generator.rst [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+
+Generating Telemetry XML Files
+-------------------------------
+
+This code block generates telemetry XML files for the packets defined in the
+`packets` dictionary. The telemetry XML files are saved in the directory specified
+by the `packet_definition_path` variable.
+
```
Comment (maxinelasp) [inline docs/source/reference/tools/xtce_generator.rst]: Is this still a WIP? It ends kind of abruptly. This section also doesn't feel very necessary - I would just move the last sentence up to the previous section (unless you're planning on adding a lot more.)

File: docs/source/reference/tools/xtce_generator.rst [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+
+
```
Comment (maxinelasp) [inline docs/source/reference/tools/xtce_generator.rst]: You can look at the other documentation for an example on how to add this file to the `toctree` in the reference index. You should also make a `tools.rst` file to act as a landing page for this tools directory - it doesn't need to include anything except another toctree, but it will act as the top-level page for any other documentation we add under tools.

**Review by maxinelasp — CHANGES_REQUESTED** [review]
The page looks good overall, I added a few suggestions. Only real changes needed are moving the XML files into another PR, and adding the document to a `toctree` so it can be discovered.

File: docs/source/reference/tools/xtce_generator.rst (hunk 2 above)
Comment (GFMoraga) [reply docs/source/reference/tools/xtce_generator.rst]: I like this!

File: docs/source/reference/tools/xtce_generator.rst (hunk 10 above)
Comment (GFMoraga) [reply docs/source/reference/tools/xtce_generator.rst]: I asked about this in the slack thread.

File: imap_processing/codice/packet_definitions/P_COD_HI_INSTRUMENT_COUNTERS.xml (hunk 1 above)
Comment (GFMoraga) [reply imap_processing/codice/packet_definitions/P_COD_HI_INSTRUMENT_COUNTERS.xml]: @maxinelasp how do I take these "out" of this commit?

File: docs/source/development/tools/xtce_generator.rst [hunk 11]
```diff
@@ -15,31 +23,29 @@ called `xtce_generator_yourinstrument.py`.
...
+----------------
 
-- Python installed (version 3.9 or higher)
-- The required Python packages installed:
-    - pathlib
-    - tools.xtce_generation.telemetry_generator
+Generating XTCEs is only done whenever packet definitions get updated, and thus it
+is not a part of the main processing package. To use it there are a few extra
+dependencies like `pandas` that you can install with
+    poetry install --extras tools
```
Comment (maxinelasp) [inline docs/source/development/tools/xtce_generator.rst:35]: Can you make this either a code block (with `::`) or make it monospaced (with two tick marks ``)

File: docs/source/development/tools/xtce_generator.rst [hunk 12]
```diff
@@ -15,31 +23,29 @@ called `xtce_generator_yourinstrument.py`.
...
+Before you Start
+----------------
 
-- Python installed (version 3.9 or higher)
-- The required Python packages installed:
-    - pathlib
-    - tools.xtce_generation.telemetry_generator
+Generating XTCEs is only done whenever packet definitions get updated, and thus it
+is not a part of the main processing package. To use it there are a few extra
+dependencies like `pandas` that you can install with
```
Comment (maxinelasp) [inline docs/source/development/tools/xtce_generator.rst]:
```suggestion
dependencies like ``pandas`` that you can install with
```

sphinx calls for two tick marks

File: docs/source/development/tools/xtce_generator.rst [hunk 13]
```diff
@@ -69,9 +78,9 @@ Follow the format below.
 Generating Telemetry XML Files
 -------------------------------
 
-This code block generates telemetry XML files for the packets defined in the
-`packets` dictionary. The telemetry XML files are saved in the directory specified
-by the `packet_definition_path` variable.
+Once you have your xtce processing file defined, you can run it with the
+following command:
+    python xtce_generator_instrument_name.py
```
Comment (maxinelasp) [inline docs/source/development/tools/xtce_generator.rst]: Put this in a code block as well

Comment (maxinelasp) [reply docs/source/development/tools/xtce_generator.rst]: Both of those code blocks also throw a warning when I ran `sphinx-build . build`:

```
/Users/maha7656/Documents/code/imap/imap_processing/docs/source/development/tools/xtce_generator.rst:32: ERROR: Unexpected indentation.
/Users/maha7656/Documents/code/imap/imap_processing/docs/source/development/tools/xtce_generator.rst:83: ERROR: Unexpected indentation.
```
So make sure you fix that before merging

**Review by maxinelasp — DISMISSED** [review]
Looks a lot better, nice! Just those two code blocks to update from me, and please take the xml files out of the review.

**Review by GFMoraga — COMMENTED** [review]
I just deleted the xtce files  and will readd in new PR.

**Review by tech3371 — APPROVED** [review]
looks good

File: imap_processing/codice/packet_definitions/P_COD_HI_INSTRUMENT_COUNTERS.xml (hunk 1 above)
Comment (tech3371) [reply imap_processing/codice/packet_definitions/P_COD_HI_INSTRUMENT_COUNTERS.xml]: I know that we can do that through git reset soft command but I will let @maxinelasp or @greglucas chime in because I struggle with those kinds of things myself.

**Comment by GFMoraga** [comment]
I need to take a look because I think I messed up my whole setup....ugh
________________________________

File: imap_processing/codice/packet_definitions/P_COD_HI_INSTRUMENT_COUNTERS.xml (hunk 1 above)
Comment (greglucas) [reply imap_processing/codice/packet_definitions/P_COD_HI_INSTRUMENT_COUNTERS.xml]: I think you're still OK here Gabriel, but it won't be super easy/obvious how to update it either. I _think_ what happened is:
1. You updated the XML generator
2. You then ran the generator and updated the codice XML files
3. You committed **both** 1 and 2
4. You got feedback to remove 2 and tried to `git rm *.xml` in that directory
5. Now those files are _removed_ and thus have been updated in the opposite direction and still show changes

If that is right, then I think we want to just get back to point 1 above and commit that update?

The way I would approach that is by trying to get back to the original work you had and then modifying that commit. Here is a list of commands that I ran to accomplish this and push to my remote so you can see what it'd look like:
https://github.com/IMAP-Science-Operations-Center/imap_processing/compare/dev...greglucas:imap_processing:gml-xtce-update-docs

Note that during the rebase, I changed the first and third commit to "edit" from "pick" so that I could stop and modify those commits and remove the files from staging. I don't think that stuff is captured here, so these commands won't be verbatim what you need. Also notice the liberal use of `git status` and `git log` so I can see where I am in the process and what the current commit has in it.

```bash
10088  gh pr checkout 163
10089  git status
10090  git log
10091  git checkout -b gml-xtce-update-docs
10092  git fetch upstream
10093  git rebase -i upstream/dev
10094  git status
10095  git reset --soft HEAD~1
10096  git status
10097  git restore --staged imap_processing/codice/packet_definitions/*.xml
10098  git status
10099  git checkout imap_processing/codice/packet_definitions/*.xml
10100  git status
10101  git log
10102  git commit
10103  git rebase --continue
10104  git status
10105  git restore --staged imap_processing/codice/packet_definitions/*.xml
10106  git status
10107  git commit
10108  git status
10109  git rebase --continue
10110  git status
10111  git push -u origin gml-xtce-update-docs
10112  git log
```

**Review by bourque — APPROVED** [review]
Nice work @GFMoraga! Looks good!

File: docs/source/development/tools/xtce_generator.rst [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+Follow the format below.
+
+.. code::
+
+    packets = {
+        # Define packet names and associated Application IDs (apId)
+        "your_packet_A": ####,
+        "your_packet_B": ####,
+        # ... (other packet definitions)
+    }
```
Comment (maxinelasp) [inline docs/source/development/tools/xtce_generator.rst:74]:
```suggestion
    }
    
```

Warning about a blank line

**Comment by GFMoraga** [comment]
@greglucas I rebased! Thank you for the command guidance.

**Review by maxinelasp — APPROVED** [review]
This looks good to me! Nice job. There is one warning to fix, I put in a comment so you can just click the button to update it.

File: docs/source/development/tools/index.rst [hunk 15]
```diff
@@ -0,0 +1,4 @@
+.. toctree::
```
Comment (greglucas) [inline docs/source/development/tools/index.rst:4]: Looking at the rendered docs: https://imap-processing--163.org.readthedocs.build/en/163/development/tools/xtce_generator.html

It looks like this page is showing up as `<no-name>` in the nav bar.

I wonder if this needs a title as well so it will show up with nested items like the style-guide entry?

```suggestion
Tools
=====

.. toctree::
```

**Review by greglucas — APPROVED** [review]
Nice work! I'm glad you were able to parse the cryptic rebase, it is never easy to recover from things like that.

**Comment by bourque** [comment]
@all-contributors please add @GFMoraga for documentation
